### ISSUE TITLE
Coroutine APIs – why Flow<T>, and not StateFlow<T>?

### ISSUE THREAD

**Issue by bvitaliyg** [russhwolf/multiplatform-settings#156]
`ObservableSettings` provide some APIs to access the settings as a `Flow<T>`.
If you want to consume it as a state (to render in UI, for example) it has to be converted to a `StateFlow<T>`:

```
val checkedState: Flow<Int> = observableSettings.intFlow("key", defaultValue = 0)
val stateFlow: StateFlow<Int> = checkedState.stateIn(coroutineScope, SharingStarted.Eagerly, initialValue = 0)

@Composable
fun MyUi(state: StateFlow<Int>) {
    val value by state.collectAsState()
    Text(text = value.toString())
}
```

Notice that you need to specify the defaultValue twice – for `intFlow()` and for `stateIn()`. This creates 2 sources for truth and can lead to potential bugs in the future.

Can we change the APIs of non-nullable flows to return `StateFlow<T>` instead of `Flow<T>`?

**Comment by russhwolf** [comment]
Setting the default value as the initial value might not always be correct. Suppose a value is saved at your key, but the first read from disk isn't immediate. Then your stateflow will return the default value followed by the on-disk value, even though a value had always been present. That might be ok for some use-cases, but I'm not convinced it's correct behavior all the time.

I could maybe see doing separate StateFlow APIs, and then you're opting into that behavior. But first I'd like to hear from more people that this is something they're looking for.

**Comment by CaelumF** [comment]
how about a suspending stateflow that suspends until the on-disk value is read? then the default is guaranteed to only be returned in the absence of an on disk value?

**Comment by bvitaliyg** [comment]
@russhwolf that's why I think reading initial settings from the disk should be always blocking – it's a critical component for the app's logic, and the resulting delay is very small in 99% of the cases – but this issue is not about that.

I think the StateFlows might be created as alternative APIs to the existing Flows. If developer utilises this API, it's clear that the developer understands the contract and responsibility coming with it.

It's not possible to create StateFlow without providing some initial value. And it's part of its logic to emit the current value first – so, in most of the cases it's clear for the developer that the 1st proceed value will be whatever is set there already by default.

Alternatively, we can create an OptIn annotation to explicitly ask the acknowledgment. What do you think?

**Comment by russhwolf** [comment]
I'm feeling more open to this as I think about it more, but not ready to commit to it yet. Biggest thing I want to be careful of is keeping consistency between `ObservableSettings` and `FlowSettings`. Here's a braindump of some of my thinking.

With `ObservableSettings`, the flow extensions are already grabbing the current value immediately and then installing a listener to pick up future updates. This very naturally makes sense as a `StateFlow`, except for the possibility that the initial read is blocking. (This now has me thinking maybe that initial read should happen inside a `withContext()` and you should pass a `CoroutineContext` into all the existing flow extensions. Hmm.) But `FlowSettings` doesn't have a natural way to do the same thing without suspending.

Without any change to the library, for `ObservableSettings` you could pretty easily define your own extensions like this:
```kotlin
fun ObservableSettings.getIntStateFlow(
    key: String,
    defaultValue: Int,
    coroutineScope: CoroutineScope
): StateFlow<Int> {
    val flow = getIntFlow(key, defaultValue)
    val initialValue = getInt(key, defaultValue)
    return flow.stateIn(coroutineScope, SharingStarted.Eagerly, initialValue)
}
```
This is a small improvement on your initial code, because it will never send the default value if there's already something stored, which was my biggest initial objection. It isn't perfect because `getIntFlow()` will make the same `getInt()` call internally, but the repeat emission won't be visible downstream because `StateFlow` discards repeat values. It's probably good enough if you want something you can put in your code today.

To make this better, use a function like this, based on the existing flow extension internals

```kotlin
inline fun <T> ObservableSettings.createStateFlow(
    coroutineScope: CoroutineScope,
    key: String,
    defaultValue: T,
    crossinline getter: Settings.(String, T) -> T,
    crossinline addListener: ObservableSettings.(String, T, (T) -> Unit) -> SettingsListener
): StateFlow<T> = callbackFlow {
    val listener = addListener(key, defaultValue) {
        trySend(it)
    }
    awaitClose {
        listener.deactivate()
    }
}.stateIn(coroutineScope, SharingStarted.Eagerly, getter(key, defaultValue))

fun ObservableSettings.getIntStateFlow(
    coroutineScope: CoroutineScope,
    key: String,
    defaultValue: Int
): StateFlow<Int> =
    createStateFlow(coroutineScope, key, defaultValue, Settings::getInt, ObservableSettings::addIntListener)
```
One thing I'm still uncertain about here is whether there are different values people might want to use rather than `SharingStarted.Eagerly`. Maybe that becomes an input parameter as well, but now the syntax is getting more complicated. But I can definitely see where this could make sense as part of the library.

But with `FlowSettings`, you don't have the option of a synchronous getter. You can use the suspending version of `stateIn()`:

```kotlin
public suspend fun FlowSettings.getIntStateFlow(
    key: String,
    defaultValue: Int,
    coroutineScope: CoroutineScope
): StateFlow<Int> {
    val flow = getIntFlow(key, defaultValue)
    return flow.stateIn(coroutineScope)
}
```
but I don't know how useful that is in practice. Are there patterns people are using for reading directly from DataStore into Compose, since that would have the same suspending issue? Or is a suspending getter for a `StateFlow` not helpful at all for your use-case? 

Incidentally, if you are ok with suspending, you could do the same thing for `ObservableSettings` and never need to worry about `SharingStarted` or `initialValue`. That would be really nice from an API consistency standpoint, but that doesn't matter if it doesn't help you do anything useful. But maybe there are non-compose use-cases where this would be good enough.

Also, if suspending is ok and all we need to do is add `.stateIn(coroutineScope)`, then it might be good enough to just do that in application code and leave the library alone.

One other caution:
> If developer utilises this API, it's clear that the developer understands the contract and responsibility coming with it.
  
I don't agree here. Developers often misunderstand APIs if they're not thoughtfully constructed. And they often don't read documentation and make wrong assumptions. I don't claim that I always know the best way to design an API, but that's why I'd rather hear from more people about what they need before committing to something.

**Comment by russhwolf** [comment]
My thinking is changing on this. The more I think about `public suspend fun FlowSettings.getIntStateFlow(): StateFlow<Int>`, the weirder it feels. There's not really a purpose to `StateFlow` in the context of `FlowSettings`, because everything in the API is suspending. So the API consistency concern I started with doesn't really matter, because `StateFlow` mostly only makes sense for `ObservableSettings` and not for `FlowSettings`. 

Since we can cast when converting to `FlowSettings`, it probably does make sense to just change the existing `getIntFlow()` and friends to return `StateFlow`. When we convert back to `ObservableSettings` via `toBlockingObservableSettings()`, we can wrap in `runBlocking()` to get the initial state since we're already doing that for other reads and it's explicitly a blocking API.

I've started doing some prep for version 1.2.0 and I'm leaning toward making this change for that release. I still haven't thought very deeply about `SharingStarted`, so how to handle that is still an open question in my mind.

**Comment by bvitaliyg** [comment]
Hey, and thank you for considering this proposal.

I think there are two scenarios possible:

1. The app can wait for the setting to load (if it's not super-critical for the app's correct initialisation). In this case, I don't think there's a need to provide something like `suspend fun getMySetting() : StateFlow<Setting>`. Just providing a regular `Flow` without a `suspend` modifier should be enough: async waiting can be achieved using regular Flow operators.

3. The setting is critical for the app's initial load. In this case, providing a `StateFlow` in a non-suspend manner is very important. This can be achieved by using a fallback (as I initially proposed) or by reading the setting in a blocking way to obtain the initial value. Each of these behaviours has its drawbacks: fallback might not be a perfect solution if the value is not expected to be changed during the app's startup, and the blocking read can take a long time if we're reading not a simple type but a complex object. Ideally, I'd go for both: if I had to choose, I think the fallback is a more universal option.

**Comment by russhwolf** [comment]
Now released in 1.2 as alternate APIs in addition to the flow extensions. This ended up being the easiest way to deal with the fact that `StateFlow` needs a coroutine scope while the raw `Flow`s don't.
